### PR TITLE
Fix stream input handling

### PR DESCRIPTION
**Pull request by danielrainer**
- Handle multiple newline-separated commands read in the same invocation of stream::on_stream_data.
- Fix memory safety issues:
    - Initialize newly allocated memory to zero.
    - Ensure that buffer for data from stdin is always \0-terminated.
- Limit lifetime of input data to functions handling the input. This keeps the input buffer size from increasing beyond the largest one-time input size (plus the linear overhead from the way allocations are increased).

### CODE REVIEW

**Comment by Ferdi265** [comment]
Hi! Thanks for the PR!

The stream string handling was way overdue for a refactoring, since it contains most of the ugly allocation and memory management complications in wl-mirror. Thanks for cleaning that up.

I will probably review this later this weekend, since I don't have time at the moment.

**Comment by Ferdi265** [comment]
> - Handle multiple newline-separated commands read in the same invocation of stream::on_stream_data.

IIRC, I actually intended for that to work with the original code as well, which is why `on_stream_data()` `memove()`s the rest data back to the front after processing, but it looks like that was broken because it then did `break;` instead of processing the rest.

Your implementation, on first glance, looks like it fails to handle when `read()` returns a partial line, which is zeroed by the call to `bzero()` at the end of the function and never processed, instead of being processed together with the next call to `read()`. One solution would be to only zero the back part of the buffer (the part between indices len and cap) when exiting the function.

I'm gonna do a full review with testing later, this is just what I saw on a first look at the PR from my phone.

**Comment by danielrainer** [comment]
> Your implementation, on first glance, looks like it fails to handle when `read()` returns a partial line, which is zeroed by the call to `bzero()` at the end of the function and never processed, instead of being processed together with the next call to `read()`. One solution would be to only zero the back part of the buffer (the part between indices len and cap) when exiting the function.

Good point, I addressed that in e65e84ed3da6c0b7bc488186ffea41498bf17a4d.
Thank you for the feedback!

**Comment by Ferdi265** [comment]
Finally had time for a full review:

TL;DR: I like your overall direction, but I would do it a little differently, mostly formatting / wording and standard API usage. I created a fixup commit and pushed it onto your branch.

- header comment for `line`/`line_len`/`line_cap`:
    - great idea to document this, though I would word it slightly differently.
    - As you suggested further down, I'm also in favor of renaming this. I chose `input`/`input_len`/`input_cap`, since it's not a line, just a buffer for unstructured input.
- header comment for `args`/`args_len`/`args_cap`:
    - similar, I added a more verbose comment
- `args_push()`/`line_reserve()`:
    - `bzero()` is deprecated in POSIX, so use `memset()` instead.
- `on_stream_data()`:
    - adding a terminating null byte to the  `read()` is ok, better be safe than sorry
    - good fix for iterating through the line, and only moving the partial line back to the front at the very end is also good, since it's much less error-prone.
    - you accidently removed the `if (line[i] == '\0') line[i] = ' ';` though, which was deliberate, added that back and documented it in the header comment for the loop
    - moving the last partial line back to the front doesn't have to be this complicated and shouldn't add another allocation. I replaced this with a `memmove()` and a `memset()`

**Comment by Ferdi265** [comment]
Renaming `line` and co to `input` done in 5e57a0f.

**Comment by danielrainer** [comment]
Great, these changes make a lot of sense.

Regarding the `\0` bytes in the input I was wondering where they would come from. Are there tools using these for argument separation instead of spaces?

**Comment by Ferdi265** [comment]
> Great, these changes make a lot of sense.
> 
> Regarding the `\0` bytes in the input I was wondering where they would come from. Are there tools using these for argument separation instead of spaces?

usually there aren't any null bytes in the input, but some UNIX coreutils have options to output as `\0` separated instead of line/space separated and I would rather do something sensible with null bytes rather than silently truncate the option string at them.
